### PR TITLE
Handler config merge without constructor magic

### PR DESCRIPTION
**Pull request by kevinazhu**
This should be a working implementation of the server which utilized both the new config format and the new RequestHandler Init function.

1. Moves a lot of parsing from NginxConfig to Server to reduce unnecessarily complex variables.
2. Implements new Config format as specified by class API
3. Converted RequestHandlers to now use Init
4. Updates test files to utilize new config and RequestHandler initialization.

CreateByName function is still not implemented because of troubles getting it working, but the standardized API should be implemented at this point.

### CODE REVIEW

**Comment by jteo1** [comment]
looks good, perhaps lets ask on piazza about getting the constructor magic working

File: request_handler.h [hunk 1]
```diff
@@ -14,12 +15,47 @@ class RequestHandler {
...
+	// Handles an HTTP request, and generates a response. Returns a response code
+	// indicating success or failure condition. If ResponseCode is not OK, the
+	// contents of the response object are undefined, and the server will return
+	// HTTP code 500.
 	virtual RequestHandler::Status HandleRequest(const Request& request, Response* response) const = 0;
 };
 
 // map of k,v pairs, where k = path and v = pointer to request handler
-typedef std::map<const std::string, std::unique_ptr<RequestHandler>> HandlerContainer;
+typedef std::map<const std::string, RequestHandler*> HandlerContainer;
```
Comment (jteo1) [inline request_handler.h]: Wait, since we changed this to RequestHandler*, we need to make a destructor to clean up the memory allocation

Comment (jteo1) [inline request_handler.h]: As a note, this was changed to match the common API requiring CreateByName to return a RequestHandler*

File: server.cc [hunk 2]
```diff
@@ -2,38 +2,26 @@
...
 #include "echo_handler.h"
 #include "file_handler.h"
+#include "not_found_handler.h"
 
 using boost::asio::ip::tcp;
 
 Server* Server::MakeServer(boost::asio::io_service& io_service, NginxConfig& out_config)
 {
+	// generate request handlers
+	auto handlers = new HandlerContainer();
```
Comment (jteo1) [inline server.cc]: Server class will need a destructor to handle cleaning this.
Also why did we make this auto again?

File: server.cc [hunk 3]
```diff
@@ -73,3 +61,86 @@ void Server::handle_accept(Connection* new_connection, const boost::system::erro
...
+					return false;
+				}
+
+				port = parsed_port;
+				port_found = true;
+			}
+
+			//default handler
+			else if(statement->tokens_[0] == "default" && statement->child_block_ != nullptr) {
+				RequestHandler* handler = RequestHandler::CreateByName(statement->tokens_[1]);
```
Comment (jteo1) [inline server.cc]: I think this was left in, we dont have CreateByName working yet

File: request_handler.h (hunk 1 above)
Comment (kevinazhu) [inline request_handler.h]: Changed to shared_ptr. CreateByName does not matter for common API, so changed to return shared_ptr as well.

**Comment by dstephan1** [comment]
I think its failing the CI.

**Comment by dstephan1** [comment]
How to make the CreateByName work:
need this: REGISTER_REQUEST_HANDLER(FileHandler) in each header

File: request_handler.h [hunk 4]
```diff
@@ -14,12 +15,47 @@ class RequestHandler {
...
+	// indicating success or failure condition. If ResponseCode is not OK, the
+	// contents of the response object are undefined, and the server will return
+	// HTTP code 500.
 	virtual RequestHandler::Status HandleRequest(const Request& request, Response* response) const = 0;
 };
 
 // map of k,v pairs, where k = path and v = pointer to request handler
-typedef std::map<const std::string, std::unique_ptr<RequestHandler>> HandlerContainer;
+typedef std::map<const std::string, std::shared_ptr<RequestHandler>> HandlerContainer;
+
```
Comment (dstephan1) [inline request_handler.h]: why the change to shared_ptr?

File: request_handler.h [hunk 5]
```diff
@@ -14,12 +15,47 @@ class RequestHandler {
...
-	// requestData is the full http request
-	// request is the parsed header
+	// Initializes the handler. Returns a response code indicating success or
+	// failure condition.
+	// uri_prefix is the value in the config file that this handler will run for.
+	// config is the contents of the child block for this handler ONLY.
+	virtual Status Init(const std::string& uri_prefix, const NginxConfig& config) = 0;
+
+	// Creation Magic by name of handler as specified in config
+	static std::shared_ptr<RequestHandler> CreateByName(const std::string& type);
```
Comment (dstephan1) [inline request_handler.h]: Why shared_ptr? It can probably be a raw pointer. If you want it can be a unique pointer.

File: file_handler.h [hunk 6]
```diff
@@ -10,14 +10,11 @@
...
-	// data is the full http request
-	// request is the parsed request
+	RequestHandler::Status Init(const std::string& uri_prefix, const NginxConfig& config);
 	virtual RequestHandler::Status HandleRequest(const Request& request, Response* response) const;
 
 private:
 	enum { max_length = 8192 };
-	static std::unordered_map<std::string,std::string> content_mappings;
 	std::string directory_;
 };
```
Comment (dstephan1) [inline file_handler.h]: You need this 
REGISTER_REQUEST_HANDLER(FileHandler)

File: request_handler.h (hunk 1 above)
Comment (kevinazhu) [inline request_handler.h]: Decided to change back to regular pointer since it doesn't matter with a lifetime of the whole server.
